content/blog/2022-05-19-A_Type_and_Effect_System_for_Deterministic_Parallel_Java/index.md

```diff
@@ -0,0 +1,123 @@
+
+### Supporting Concurrency in Arrays
+
+In order to allow safe concurrency in arrays, the authors define two novel capabilities: (1) **index-parameterized arrays** and (2) **subarrays**.
+
+##### Index-Parameterized Arrays
+
+Index-Parameterized Arrays allow to describe concurrent operations on different elements of an array. The problem seems to be trivial for arrays of primitive types, however, it is way harder for arrays of **mutable references**. The idea behind index-parameterized arrays is simple: we need to parameterize each object reference in the array by the array index's integer expression. For example, the aforementioned ```TreeNode``` class should be parameterized as follows ```TreeNode<Root:[10]>``` for the element ```10``` in the array. The compiler then will check (with a symbolic integer expression solver) that all array accesses in the program going to the element parameterized by expression [i] happen through the array element ```i```. If this holds, we get formally proved safe concurrent access to all the distinct elements in the array, even with mutable references.
+
+##### Subarrays
+
+In addition to the aforementioned Index-Parameterized Arrays, DPJ also provides the functionality to express concurrent operations on sub-arrays. This is very useful for divide-and-conquer algorithms. Here, each array can be partitioned with the new ```DPJPartition``` abstraction that represents a distinct memory region. ```DPJPartition```s are parameterized by two integer expressions: the starting position and the length of the partition. Similarly to the above Index-Parameterized Arrays, compiler checks that all distinct concurrent memory accesses go to different partitions.
+
+### Commutativity Annotations
+
+This is a simple feature that attempts to generalize `read`/`write` effects by explicitly specifying what methods are commutative and therefore can be called concurrently. For example, 
+```
+class IntSet<region P> {
+  void add(int x) writes P { ... }
+  add commuteswith add;
+}
+```
+defines that two invocations of `add` are atomic and either order of invocations produces the same result. The commutativity property is not getting checked by the compiler and should be verified independently with formal analysis.
+
+### Evaluation
+
+The authors evaluate their type and effect system on a set of benchmarks and show that it indeed allows to express concurrency that provides near-ideal linear scaling of performance with the number of executors. Some algorithms do not scale linearly due to certain limitations of DPJ such as impossibility of re-shuffling for array elements. Additionally, while these techniques allow some realistic concurrent algorithms to be expressed in this language, threads which write to the same location in memory must make use of the commutativity annotations, and the correctness of these annotations must be verified by the programmer.
```

Comment:
This is one place where some specifics could be helpful. There is absolutely no need to recapitulate the entire evaluation, of course! But mentioning in a little more detail why _one_ benchmark scaled more poorly in DPJ could help make a discrete point about its strengths vs. weaknesses.